src/backend/distributed/commands/foreign_server.c

```diff
@@ -0,0 +1,309 @@
+	ObjectAddressSet(address, ForeignServerRelationId, server->serverid);
+
+	/* filter distributed servers */
+	if (!ShouldPropagateObject(&address))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	char *sql = DeparseTreeNode(node);
+
+	/* to prevent recursion with mx we disable ddl propagation */
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) sql,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * PreprocessDropForeignServerStmt is called during the planning phase for
+ * DROP SERVER.
+ */
+List *
+PreprocessDropForeignServerStmt(Node *node, const char *queryString,
+								ProcessUtilityContext processUtilityContext)
+{
+	DropStmt *stmt = castNode(DropStmt, node);
+	Assert(stmt->removeType == OBJECT_FOREIGN_SERVER);
+
+	List *allServerNamesToDrop = stmt->objects;
+	List *distributedServerAddresses = NIL;
+	List *distributedServerNames = NIL;
+	Value *serverValue = NULL;
+	foreach_ptr(serverValue, stmt->objects)
+	{
+		char *serverString = strVal(serverValue);
+		ForeignServer *server = GetForeignServerByName(serverString, false);
+
+		ObjectAddress address = { 0 };
+		ObjectAddressSet(address, ForeignServerRelationId, server->serverid);
+
+		/* filter distributed servers */
+		if (IsObjectDistributed(&address))
+		{
+			distributedServerAddresses = lappend(distributedServerAddresses, &address);
+			distributedServerNames = lappend(distributedServerNames, serverValue);
+		}
+	}
+
+	if (list_length(distributedServerNames) <= 0)
+	{
+		return NIL;
+	}
+
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	/* unmark each distributed server */
+	ObjectAddress *address = NULL;
+	foreach_ptr(address, distributedServerAddresses)
+	{
+		UnmarkObjectDistributed(address);
+	}
+
+	/*
+	 * Temporary swap the lists of objects to delete with the distributed
+	 * objects and deparse to an sql statement for the workers.
+	 * Then switch back to allServerNamesToDrop to drop all specified
+	 * servers in coordinator after PreprocessDropForeignServerStmt completes
+	 * its execution.
+	 */
+	stmt->objects = distributedServerNames;
+	const char *deparsedStmt = DeparseTreeNode((Node *) stmt);
+	stmt->objects = allServerNamesToDrop;
+
+	/*
+	 * To prevent recursive propagation in mx architecture, we disable ddl
+	 * propagation before sending the command to workers.
+	 */
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) deparsedStmt,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * PostprocessCreateForeignServerStmt is called after a CREATE SERVER command has
+ * been executed by standard process utility.
+ */
+List *
+PostprocessCreateForeignServerStmt(Node *node, const char *queryString)
+{
+	bool missingOk = false;
+	ObjectAddress typeAddress = GetObjectAddressFromParseTree(node, missingOk);
+	EnsureDependenciesExistOnAllNodes(&typeAddress);
```

Comment:
What kind of dependency a foreign server might have? Do we have a test for that?

Reply:
So, I think a server can depend to an extension, and an extension can depend on a schema? So, such a test might be nice:
```sQL
cREATE SCHEMa sc1;
CREATE EXTENSION postgres_fdw WITH SCHEMA sc1;
CREATE SERVER .... 

-- make sure that schema is also distributed
```

Reply:
Added, thanks!